node/src/main/scala/org/bitcoins/node/Node.scala

```diff
@@ -159,30 +187,37 @@ trait Node extends NodeApi with ChainQueryApi with P2PLogger {
     val chainApiF = startConfsF.flatMap(_ => chainApiFromDb())
 
     val startNodeF = {
-      val isInitializedFs = peerMsgSenders.indices.map { idx =>
-        peerMsgSenders(idx).connect()
-        val isInitializedF = for {
-          _ <- AsyncUtil.retryUntilSatisfiedF(() => isInitialized(idx),
-                                              maxTries = 200,
-                                              interval = 250.millis)
-        } yield ()
-        isInitializedF.failed.foreach(err =>
-          logger.error(
-            s"Failed to connect with peer=${peers(idx)} with err=$err"))
-        isInitializedF.map { _ =>
-          nodeAppConfig.nodeType match {
-            case NodeType.NeutrinoNode => {
-              if (peerServices(peers(idx)).nodeCompactFilters) {
-                logger.info(s"Our peer=${peers(idx)} has been initialized")
-              } else {
-                logger.info(
-                  s"Our peer=${peers(idx)} does not support compact filters. Disconnecting.")
-                peerMsgSenders(idx).disconnect()
+      logger.info(peers)
+      val isInitializedFs = peers.map { peer =>
+        peerData(peer).peerMessageSender.connect()
+        val isInitializedF =
+          for {
+            _ <- AsyncUtil
+              .retryUntilSatisfiedF(
+                () => peerData(peer).peerMessageSender.isInitialized(),
+                maxTries = 50,
+                interval = 250.millis)
+              .recover { case NonFatal(_) =>
+                logger.info(s"Failed to initialize $peer")
+                peerData(peer).peerMessageSender.disconnect()
+                removePeer(peer)
               }
+          } yield ()
+        isInitializedF.map { _ =>
+          if (peerData.contains(peer)) {
```

Comment:
is is possible that `peerData` _doesn't_ contain the peer?

Reply:
Yes, there were peers we could connect to but not initialize, in which case they were removed, so it might happen that this map would be called corresponding to a removed peer.